// @holochain-open-dev/profiles/dist/profiles-store.js
import { LazyHoloHashMap, slice } from "@holochain-open-dev/utils";
import { asyncDeriveStore, asyncReadable, lazyLoadAndPoll, joinAsyncMap, lazyLoad, manualReloadStore, } from "@holochain-open-dev/stores";
import { defaultConfig } from "./config.js";
export class ProfilesStore {
    constructor(client, config = {}) {
        this.client = client;
        /**
         * Fetches all the agents that have created a profile in the DHT
         */
        this.agentsWithProfile = lazyLoadAndPoll(() => this.client.getAgentsWithProfile(), 1000);
        /**
         * Fetches the profiles for all agents in the DHT
         *
         * This will get slower as the number of agents in the DHT increases
         */
        this.allProfiles = asyncDeriveStore(this.agentsWithProfile, (agents) => this.agentsProfiles(agents));
        // Fetches the profile for the given agent
        this.profiles = new LazyHoloHashMap((agent) => asyncReadable(async (set) => {
            const profile = await this.client.getAgentProfile(agent);
            set(profile);
            return this.client.onSignal((signal) => {
                if (this.client.client.myPubKey.toString() !== agent.toString())
                    return;
                if (!(signal.type === "EntryCreated" || signal.type === "EntryUpdated"))
                    return;
                set(signal.app_entry);
            });
        }));
        // Fetches your profile
        // TODO: change the manual implementation when signals are working consistently
        this.myProfile = manualReloadStore(async () => this.client.getAgentProfile(this.client.client.myPubKey));
        this.config = { ...defaultConfig, ...config };
    }
    // Fetches the profiles for the given agents
    agentsProfiles(agents) {
        return joinAsyncMap(slice(this.profiles, agents));
    }
    searchProfiles(searchFilter) {
        return asyncDeriveStore(lazyLoad(async () => this.client.searchAgents(searchFilter)), (agents) => this.agentsProfiles(agents));
    }
}
